# **Exploratory Data Analisys: Companies**

## **Initial Setup**

### Install Packages

In [18]:
%pip install pandas -q
%pip install plotly -q

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### Import libs

In [19]:
import os
import itertools
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp
import warnings
warnings.filterwarnings('ignore')

### Pandas Config

In [20]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [21]:
file_path_book = str(Path(os.getcwd()).parent.parent.parent / "data/book")

## Macroeconomic

### Load data

In [22]:
df_macroeconomic_book = pd.read_csv(file_path_book + "/macroeconomic_book.csv")
df_macroeconomic_book.head(5)

date  selic  confidence       pib  incc  ipca   dolar  \
0  2019-01-31    6.5      128.64  578214.5  0.49  3.78  3.6513   
1  2019-02-28    6.5      139.39  576089.7  0.09  3.89  3.7379   
2  2019-03-31    6.5      125.53  601749.8  0.31  4.58  3.8961   
3  2019-04-30    6.5      121.71  612918.4  0.38  4.94  3.9447   
4  2019-05-31    6.5      117.01  615304.9  0.03  4.66  3.9401   

   monthly_inflation  gdp_growth  dollar_growth  real_interest_rate  \
0          -0.111948   -0.015756       0.016459                2.72   
1           0.029101   -0.003675       0.023718                2.61   
2           0.177378    0.044542       0.042323                1.92   
3           0.078603    0.018560       0.012474                1.56   
4          -0.056680    0.003894      -0.001166                1.84   

   inflation_confidence_difference  
0                          -124.86  
1                          -135.50  
2                          -120.95  
3                          -116.77  
4                          -112.35

In [23]:
df_macroeconomic_numeric_cols = df_macroeconomic_book.select_dtypes(include=["int", "number", "float64"])
df_macroeconomic_numeric_cols.head(5)

selic  confidence       pib  incc  ipca   dolar  monthly_inflation  \
0    6.5      128.64  578214.5  0.49  3.78  3.6513          -0.111948   
1    6.5      139.39  576089.7  0.09  3.89  3.7379           0.029101   
2    6.5      125.53  601749.8  0.31  4.58  3.8961           0.177378   
3    6.5      121.71  612918.4  0.38  4.94  3.9447           0.078603   
4    6.5      117.01  615304.9  0.03  4.66  3.9401          -0.056680   

   gdp_growth  dollar_growth  real_interest_rate  \
0   -0.015756       0.016459                2.72   
1   -0.003675       0.023718                2.61   
2    0.044542       0.042323                1.92   
3    0.018560       0.012474                1.56   
4    0.003894      -0.001166                1.84   

   inflation_confidence_difference  
0                          -124.86  
1                          -135.50  
2                          -120.95  
3                          -116.77  
4                          -112.35

### Macroeconomic Analisys

#### Data Table

In [24]:
df_macroeconomic_book.shape

(56, 12)

- The dataset contains 56 rows, one day per row, and 12 columns. Each column represents a metric of the macroecomy.

In [25]:
df_macroeconomic_book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   date                             56 non-null     object 
 1   selic                            56 non-null     float64
 2   confidence                       56 non-null     float64
 3   pib                              56 non-null     float64
 4   incc                             56 non-null     float64
 5   ipca                             56 non-null     float64
 6   dolar                            56 non-null     float64
 7   monthly_inflation                56 non-null     float64
 8   gdp_growth                       56 non-null     float64
 9   dollar_growth                    56 non-null     float64
 10  real_interest_rate               56 non-null     float64
 11  inflation_confidence_difference  56 non-null     float64
dtypes: float64(11), object(1

- The dataframe df_macroeconomic_book is primarily composed of fields of type float64 to represent the indices, with the only exception being the 'date' field, which is of type object.

In [26]:
df_macroeconomic_book.describe()

selic  confidence            pib       incc       ipca      dolar  \
count  56.000000   56.000000      56.000000  56.000000  56.000000  56.000000   
mean    7.571429  114.484107  729338.171429   0.661964   5.908036   4.941593   
std     4.485511    9.230923  106504.361049   0.596372   3.036075   0.601565   
min     2.000000   96.820000  569278.800000   0.030000   1.880000   3.651300   
25%     3.750000  107.365000  636251.825000   0.217500   3.692500   4.670750   
50%     6.500000  111.995000  724215.500000   0.460000   4.630000   5.160550   
75%    13.250000  121.837500  826032.925000   0.865000   8.445000   5.403750   
max    13.750000  139.390000  927590.500000   2.280000  12.130000   5.771200   

       monthly_inflation  gdp_growth  dollar_growth  real_interest_rate  \
count          56.000000   56.000000      56.000000           56.000000   
mean            0.011175    0.008724       0.006616            1.663393   
std             0.140963    0.041070       0.045178            4.335886   
min            -0.276824   -0.107443      -0.078151           -4.740000   
25%            -0.090481   -0.011851      -0.025877           -1.335000   
50%             0.014527    0.005607       0.005248            1.070000   
75%             0.080079    0.032610       0.037589            3.205000   
max             0.318043    0.124708       0.155621           10.590000   

       inflation_confidence_difference  
count                        56.000000  
mean                       -108.576071  
std                          10.646899  
min                        -135.500000  
25%                        -117.057500  
50%                        -107.520000  
75%                         -98.902500  
max                         -91.710000

- `selic`: The average SELIC rate during the period was **7.57%**. SELIC is the reference interest rate in Brazil, which directly influences the cost of credit and can impact the behavior of investors and consumers.

- `confidence`: The average Economic Confidence was **114.48**, indicating the overall level of confidence among economic agents in the stability and growth of the economy. High confidence may suggest a positive economic environment.

- `pib`: The Gross Domestic Product (GDP) averaged **BRL 729.338.17** billion during the period. GDP is a fundamental measure of a country's economic activity, reflecting the total value of goods and services produced.

- `incc`: The average National Construction Cost Index (INCC) was **0.66**, which is related to construction costs. Variations in INCC can affect the construction industry and, by extension, the overall economy.

- `ipca`: The average Consumer Price Index for Broad Consumer Price Index (IPCA) was **5.91%**, representing the average inflation rate during the period. Inflation affects the purchasing power of the currency and the cost of living for the population.

- `dolar`: The average exchange rate was **BRL 4.94** per US dollar. Exchange rate fluctuations can impact international trade, export competitiveness, and the prices of imported goods.

- `monthly_inflation`: The average monthly inflation was **0.01%**, indicating the variation in prices over the months. Price stability is important for maintaining the purchasing power of the currency.

- `gdp_growth`: The average GDP growth was **0.01%**, reflecting the economic performance over time. Economic growth is a key indicator of a country's development.

- `dollar_growth`: The average dollar growth was **0.01%**, showing how the local currency appreciated or depreciated against the dollar during the period.

- `real_interest_rate`: The average Real Interest Rate was **1.66%**, taking inflation into account. Low real interest rates can encourage investment and consumption.

- `inflation_confidence_difference`: The average difference in confidence regarding inflation was **-108.58**, suggesting a possible lack of confidence in price stability. This can affect the consumption and investment decisions of individuals and businesses.


In [27]:
df_macroeconomic_book.nunique()

date                               56
selic                              21
confidence                         56
pib                                56
incc                               46
ipca                               55
dolar                              56
monthly_inflation                  56
gdp_growth                         56
dollar_growth                      56
real_interest_rate                 55
inflation_confidence_difference    56
dtype: int64

In [28]:
df_macroeconomic_book.isna().count()

date                               56
selic                              56
confidence                         56
pib                                56
incc                               56
ipca                               56
dolar                              56
monthly_inflation                  56
gdp_growth                         56
dollar_growth                      56
real_interest_rate                 56
inflation_confidence_difference    56
dtype: int64

- There are no missing values in the dataset.

#### Coefficient of variation

In [29]:
df_macroeconomic_book_cvs = ((df_macroeconomic_numeric_cols.std(ddof=1) / df_macroeconomic_numeric_cols.mean().sort_values(ascending=True)) * 100).sort_values(ascending=True).reset_index()
df_macroeconomic_book_cvs.columns = ['column', 'coefficient_variation']
df_macroeconomic_book_cvs['coefficient_variation'] = df_macroeconomic_book_cvs['coefficient_variation'].abs().sort_values(ascending=True)
df_macroeconomic_book_cvs.fillna(method='ffill', inplace=True)
df_macroeconomic_book_cvs = df_macroeconomic_book_cvs.sort_values(by='coefficient_variation', ascending=False)


fig = px.bar(df_macroeconomic_book_cvs, title='Coefficient of Variation (%)',y='column', x='coefficient_variation', orientation='h', color_discrete_sequence=['rgb(100, 195, 181)'],hover_name='column', height=800)
fig.update_traces(text=[f'{x:.0f}%' for x in df_macroeconomic_book_cvs['coefficient_variation']], textposition='outside')
fig.update_layout(xaxis_title='Coefficient of Variation (%)', yaxis_title='Columns', template='plotly_dark', font=dict(color='white'), xaxis={'type':'log', 'title':'Coefficient of Variation (Log Scale)'}, hovermode='y')

fig.show()

This horizontal bar graph displays the `Coefficient of Variation` for several economic indicators, indicating their relative variability.

**Coefficient of Variation Analysis**:

- `confidence`: Shows the least variability with a CV of **8%**, suggesting stable consumer confidence levels over the period.
- `inflation_confidence_difference`: Has a slightly higher CV of **10%**, which may reflect the gap between expected and realized inflation.
- `dolar`: Displays a CV of **12%**, indicating moderate variability in the dollar exchange rate.
- `pib` (GDP): Has a CV of **15%**, pointing to fluctuations in economic output.
- `ipca`: Shows higher variability with a CV of **51%**, a significant indicator of inflation volatility.
- `selic`: Exhibits a CV of **59%**, highlighting variability in the policy interest rate.
- `incc`: Shows a very high CV of **90%**, indicating substantial variability in the National Construction Cost Index.
- `real_interest_rate`: With a CV of **261%**, it suggests extreme variability in the real interest rates over the period.
- `gdp_growth`: Has an even higher CV of **471%**, pointing to the highly volatile nature of economic growth rates.
- `dolar_growth`: Shows a CV of **683%**, signaling pronounced fluctuations in the dollar growth rate.
- `monthly_inflation`: Exhibits the highest CV on the chart at **1261%**, indicating extreme volatility in monthly inflation rates.

**Interpretation**:

- The indicators with lower CVs, like `confidence`, are relatively stable, whereas those with higher CVs, such as `monthly_inflation`, show greater fluctuation and unpredictability.
- High CV values, especially in key economic metrics like `monthly_inflation` and `gdp_growth`, can suggest a challenging environment for economic forecasting and policy-making.

This chart is essential for analysts and policymakers to assess the stability and predictability of different economic measures, which is crucial for planning and decision-making.


#### Outlier Analysis

In [30]:
num_columns = len(df_macroeconomic_book)
num_rows = (num_columns + 2) // 2

subplot_titles = [str(col) for col in df_macroeconomic_book]

fig = sp.make_subplots(rows=num_rows, cols=3, subplot_titles=subplot_titles)

for i, column in enumerate(df_macroeconomic_book, start=1):
    row = (i - 1) // 3 + 1
    col = (i - 1) % 3 + 1

    trace = go.Box(y=df_macroeconomic_book[column], name=column, 
                   marker_color='lightseagreen', boxpoints='outliers', 
                   jitter=0.7, hoverinfo='y+text', 
                   text=())

    fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text='Boxplot of Numerical Variables', height=150*num_columns, showlegend=False, template='plotly_dark')
fig.show()

**Financial Rates and Inflation**

- `real_interest_rate` **(10.59)**: Presents outliers that suggest periods of exceptional real interest rates, diverging from the norm.

**Economic Indicators**

- `incc` **(1.89 ~ 2.28)**: Features outliers, indicating occasional extreme values outside the general trend of construction cost index changes.

**Confidence and Growth Metrics**

- `gdp_growth` **(-0.10, 0.11 and 0.12)**: Exhibits outliers on both ends, reflecting sporadic and unexpected shifts in the GDP growth rate.
- `dollar_growth` **(0.15)**: Outliers here indicate atypical fluctuations in the dollar growth rate, separate from the central trend.


### Histogram

In [31]:
subplot_titles = [str(col) for col in df_macroeconomic_book]
columns_per_row = 3

num_rows = len(df_macroeconomic_numeric_cols) // columns_per_row + (len(df_macroeconomic_numeric_cols) % columns_per_row > 0)
fig = sp.make_subplots(rows=num_rows, cols=columns_per_row, subplot_titles=subplot_titles)

for i, column in enumerate(df_macroeconomic_numeric_cols):
    row = i // columns_per_row + 1
    col = i % columns_per_row + 1
    
    fig.add_trace(go.Histogram(x=df_macroeconomic_numeric_cols[column], name=column, marker_color='lightseagreen'), row=row, col=col)

fig.update_layout(title='Histograms by Column', height=300 * num_rows, showlegend=False, template='plotly_dark')
fig.show()

*`Selic`*: The values range from **2.00%** to **13.75%**, with a distribution that appears to be multimodal, indicating distinct periods of monetary policy.

*`Confidence`*: Varies from **96.82** to **139.39**, with a distribution that can be approximately normal but multimodal, reflecting different periods of economic confidence.

*G`DP-PIB`*: Fluctuates between **569278.8** and **927590.5**, with an asymmetric distribution and possible clusters of values suggesting variations in economic activity.

*`INCC`*: Oscillates between **0.03** and **2.28**, with a strongly right-skewed distribution, indicating stability in construction costs with sporadic significant increases.

*`IPCA`*: Ranges from **1.88** to **12.13**, with a right-skewed distribution, suggesting periods of variable inflation, with a tendency toward higher values.

*`Dollar`*: Has values between **3.6513** and **5.7712**, with a bimodal distribution that may reflect different currency price regimes over time.

*`Monthly Inflation`*: Varies from **-0.276824** to **0.318043**, with a distribution that can be considered approximately normal but with occurrences of deflation and inflation.

*`GDP Growth`*: Presents variations from **-0.107443** to **0.124708**, with a distribution that seems to be approximately symmetric around zero, representing periods of economic contraction and expansion.

*`Dollar Growth`*: Shows a variation from **-0.078151** to **0.155621**, with a slightly right-skewed distribution, suggesting moments of more pronounced dollar appreciation.

*`Real Interest Rate`*: With a range between **-4.74** and **10.59**, the distribution is skewed with a concentration of lower values and a long tail to the positive side, indicating a variety of real monetary conditions.

*`Inflation Confidence Difference`*: Varies between **-135.50** and **-93.51**, with a distribution skewed towards negative values, reflecting moments when inflation expectations were significantly lower than overall economic confidence.

These distributions reflect the complexity of economic variables and their fluctuations over time, highlighting the importance of analyzing these indicators together to understand economic health and market confidence.


### Correlation and Dispersion

In [32]:

df_macroeconomic_numeric_cols_corr = df_macroeconomic_numeric_cols.corr()
color_scale = [[0, 'rgb(150, 245, 231)'], [0.5, 'rgb(100, 195, 181)'], [1, 'rgb(50, 145, 131)']]
fig = px.imshow(df_macroeconomic_numeric_cols_corr, text_auto=True, aspect="auto", template="plotly_dark", title="Mapa de Calor: Correlação entre as variaveis", color_continuous_scale=color_scale)
fig.show()

**Strong Negative Correlations**:

- `confidence` ↔ `inflation_confidence_difference`: **-0.963305**

**Cause and Effect**: High general confidence may lead to a smaller difference in inflationary confidence, suggesting that when people are more confident in the economy, they tend to perceive fewer inflation risks.

- `inflation_confidence_difference` ↔ `confidence`: **-0.963305**

**Cause and Effect**: A larger difference in inflationary confidence can indicate economic uncertainties, which typically reduces overall confidence in the economy.

**Strong Positive Correlations**:

- `selic` ↔ `pib`: **0.809282**

**Cause and Effect**: An increase in the Selic rate may be associated with GDP growth, possibly reflecting a monetary policy that responds to economic growth.

- `pib` ↔ `selic`: **0.809282**

**Cause and Effect**: GDP growth can lead to adjustments in the Selic rate, as part of monetary policy to control inflation and stimulate investment.

- `selic` ↔ `real_interest_rate`: **0.763599**

**Cause and Effect**: An increase in the Selic rate may correlate with an increase in the real interest rate, indicating a more restrictive monetary policy to control inflation.

- `real_interest_rate` ↔ `selic`: **0.763599**

**Cause and Effect**: High real interest rates can influence the increase of the Selic rate, as a measure to balance the economy and control inflation.

**Moderate Correlations**:

- `ipca` ↔ `inflation_confidence_difference`: **0.577951**

**Cause and Effect**: Variations in IPCA may be moderately linked to changes in inflation perception, affecting inflationary confidence.

- `inflation_confidence_difference` ↔ `ipca`: **0.577951**

**Cause and Effect**: A change in inflationary confidence may reflect changes in IPCA, influencing expectations of future inflation.

- `confidence` ↔ `real_interest_rate`: **0.508671**

**Cause and Effect**: An increase in general confidence might correlate with a rise in real interest rates, potentially reflecting optimistic economic perceptions leading to tighter monetary policy.

- `real_interest_rate` ↔ `confidence`: **0.508671**

**Cause and Effect**: Higher real interest rates could indicate growing economic confidence, which in turn can lead to an increase in investment and economic activity.

- `incc` ↔ `inflation_confidence_difference`: **0.499014**

**Cause and Effect**: Fluctuations in the INCC might relate to changes in inflationary confidence, indicating how construction costs impact inflation perceptions.

- `inflation_confidence_difference` ↔ `incc`: **0.499014**

**Cause and Effect**: A variation in inflationary confidence can be moderately influenced by changes in construction costs (INCC), reflecting the construction sector's impact on the overall economy.

- `pib` ↔ `real_interest_rate`: **0.495798**

**Cause and Effect**: The relationship between GDP and real interest rates might indicate how economic growth influences monetary policy.

- `real_interest_rate` ↔ `pib`: **0.495798**

**Cause and Effect**: Changes in real interest rates can be a response to GDP fluctuations, as part of efforts to balance economic growth and inflation.

- `pib` ↔ `ipca`: **0.487577**

**Cause and Effect**: GDP movements can be moderately correlated with IPCA fluctuations, showing how overall economic performance influences inflation rates.

- `ipca` ↔ `pib`: **0.487577**

**Cause and Effect**: Changes in IPCA may reflect the economic health represented by GDP, with inflation rates responding to economic growth or contraction.

- `dolar` ↔ `inflation_confidence_difference`: **0.451153**

**Cause and Effect**: The dollar's value might influence inflationary confidence, as currency fluctuations can impact import prices and inflation perceptions.

- `inflation_confidence_difference` ↔ `dolar`: **0.451153**

**Cause and Effect**: Changes in inflationary confidence can reflect movements in the dollar value, indicating how currency fluctuations impact economic perceptions.

- `pib` ↔ `dolar`: **0.403385**

**Cause and Effect**: The relationship between GDP and the dollar can show how economic performance influences currency value, affecting international trade and investment.

- `dolar` ↔ `pib`: **0.403385**

**Cause and Effect**: Fluctuations in the dollar's value can reflect changes in GDP, indicating the economy's strength or weakness in the global market.

- `selic` ↔ `ipca`: **0.386892**

**Cause and Effect**: The Selic rate's correlation with IPCA might show how monetary policy is adjusted in response to inflation rates.

- `ipca` ↔ `selic`: **0.386892**

**Cause and Effect**: Changes in IPCA can influence the Selic rate, as part of efforts to control inflation through monetary policy.

- `incc` ↔ `ipca`: **0.348632**

**Cause and Effect**: The INCC's relationship with IPCA could indicate how construction costs contribute to overall inflation rates.

- `ipca` ↔ `incc`: **0.348632**

**Cause and Effect**: Variations in IPCA may reflect changes in the construction sector, as indicated by movements in the INCC.

- `ipca` ↔ `dolar`: **0.312432**

**Cause and Effect**: The correlation between IPCA and the dollar value can show how currency strength influences inflation, particularly through import prices.

- `dolar` ↔ `ipca`: **0.312432**

**Cause and Effect**: Changes in the dollar value might reflect shifts in inflation rates (IPCA), indicating how external economic factors influence domestic inflation.


In [33]:
fig = px.scatter_matrix(df_macroeconomic_numeric_cols_corr, dimensions=df_macroeconomic_numeric_cols_corr.columns, title='Gráficos de Dispersão para Todas as Colunas', template="plotly_dark",)
fig.update_traces(marker=dict(color='rgb(100, 195, 181)'))

fig.update_layout(width=4000, height=3000, grid=dict(xgap=0.1, ygap=0.5))

fig.show()

**selic (Interest Rate)**

- **Strong Positive Correlation**: `selic` exhibits a **strong positive correlation** of approximately **0.80** with `pib` (GDP). This indicates that changes in the interest rate (`selic`) tend to positively influence GDP (`pib`), suggesting a significant economic relationship.

- **Moderate Positive Correlation**: There is a **moderate positive correlation** of about **0.26** between `selic` and `confidence`. This suggests that as the interest rate (`selic`) increases, confidence tends to improve, although the correlation is not very strong.

**confidence (Confidence Index)**

- **Negative Correlation**: `confidence` shows a **negative correlation** of approximately **-0.46** with `incc`. This means that as confidence decreases, the construction cost index (`incc`) tends to increase, indicating a potential inverse relationship between confidence and construction costs.

- **Strong Negative Correlation**: There is a **strong negative correlation** of around **-0.96** between `confidence` and `inflation_confidence_difference`. This indicates that as confidence drops, the inflation confidence difference tends to increase significantly.

**pib (GDP)**

- **Strong Positive Correlation**: `pib` has a **strong positive correlation** of approximately **0.81** with `selic` (Interest Rate). This implies that changes in GDP (`pib`) are positively influenced by fluctuations in the interest rate (`selic`).

- **Positive Correlation**: There is also a **positive correlation** of about **0.49** between `pib` and `inflation_confidence_difference`. This suggests that as GDP increases, the inflation confidence difference tends to rise as well.

**incc (Construction Cost Index)**

- **Strong Positive Correlation**: `incc` exhibits a **strong positive correlation** of approximately **0.50** with `inflation_confidence_difference`. This indicates that as the construction cost index (`incc`) increases, the inflation confidence difference tends to increase as well.

- **Negative Correlation**: There is a **negative correlation** of about **-0.50** between `incc` and `confidence`. As the construction cost index (`incc`) rises, confidence tends to decrease.

**ipca (Consumer Price Index)**

- **Positive Correlation**: `ipca` shows a **positive correlation** of approximately **0.58** with `inflation_confidence_difference`. This suggests that as the Consumer Price Index (`ipca`) increases, the inflation confidence difference tends to rise as well.

- **Moderate Positive Correlation**: There is a **moderate positive correlation** of about **0.31** between `ipca` and `pib`. This implies that as the Consumer Price Index (`ipca`) increases, GDP (`pib`) tends to show moderate growth.

**dolar (Exchange Rate)**
- **Positive Correlation**: `dolar` exhibits a **positive correlation** of approximately **0.45** with `inflation_confidence_difference`. This suggests that as the exchange rate (`dolar`) increases, the inflation confidence difference tends to rise as well.

- **Weak Positive Correlation**: There is a **weak positive correlation** of about **0.03** between `dolar` and `gdp_growth`. This indicates a minimal relationship between the exchange rate (`dolar`) and GDP growth (`gdp_growth`).

**monthly_inflation (Monthly Inflation Rate)**

- **Negative Correlation**: `monthly_inflation` shows a **negative correlation** of approximately **-0.38** with `inflation_confidence_difference`. As the monthly inflation rate (`monthly_inflation`) decreases, the inflation confidence difference tends to increase.

- **Moderate Positive Correlation**: There is a **moderate positive correlation** of about **0.18** between `monthly_inflation` and `gdp_growth`. This suggests that as the monthly inflation rate (`monthly_inflation`) rises, GDP growth (`gdp_growth`) tends to show moderate growth.

**gdp_growth (GDP Growth Rate)**

- **Positive Correlation**: `gdp_growth` exhibits a **positive correlation** of approximately **0.18** with `monthly_inflation`. As GDP growth (`gdp_growth`) increases, the monthly inflation rate (`monthly_inflation`) tends to rise as well.

- **Weak Positive Correlation**: There is a **weak positive correlation** of about **0.07** between `gdp_growth` and `dollar_growth`. This implies a minimal relationship between GDP growth (`gdp_growth`) and dollar growth (`dollar_growth`).

**dollar_growth (Dollar Growth Rate)**

- **Negative Correlation**: `dollar_growth` shows a **negative correlation** of approximately **-0.21** with `inflation_confidence_difference`. This suggests that as dollar growth (`dollar_growth`) decreases, the inflation confidence difference tends to increase.

- **Weak Positive Correlation**: There is a **weak positive correlation** of about **0.03** between `dollar_growth` and `gdp_growth`. This indicates a minimal relationship between dollar growth (`dollar_growth`) and GDP growth (`gdp_growth`).

**real_interest_rate (Real Interest Rate)**

- **Strong Positive Correlation**: `real_interest_rate` has a **strong positive correlation** of approximately **0.76** with `selic` (Interest Rate). This indicates that changes in the real interest rate (`real_interest_rate`) tend to positively influence the interest rate (`selic`).

- **Negative Correlation**: There is a **negative correlation** of about **-0.53** between `real_interest_rate` and `inflation_confidence_difference`. As the real interest rate (`real_interest_rate`) increases, the inflation confidence difference tends to decrease.

**inflation_confidence_difference (Inflation Confidence Difference)**

- **Strong Negative Correlation**: `inflation_confidence_difference` exhibits a **strong negative correlation** of approximately **-0.96** with `confidence`. This suggests that as the inflation confidence difference decreases, confidence tends to increase significantly.

- **Moderate Positive Correlation**: There is a **moderate positive correlation** of about **0.58** between `inflation_confidence_difference` and `ipca` (Consumer Price Index). This indicates that as the inflation confidence difference increases, the Consumer Price Index (`ipca`) tends to rise as well.

These summaries provide insights into the relationships between various economic indicators, highlighting the strengths and directions of their correlations.
